Marketplace: update category selector

raicem opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Remove the dropdown on the desktop view and show all items, even if overflowing. Added helper buttons to scroll to the right to show more.

Closes https://github.com/Automattic/woocommerce.com/issues/21591.

### How to test the changes in this Pull Request:

1. Checkout this branch and get a new build
2. Load the [In-App Marketplace](http://localhost:8888/wp-admin/admin.php?page=wc-admin&path=%2Fextensions) and move over to the Extensions tab. Confirm category selector loads. 
3. Narrow the viewport so that the category selector overflows. Confirm button to indicate scroll appears. 
4. Scroll using the helper button, when you arrive at to the **Marketing** category, the button should disappear. A new button should appear at the start position to take you back to the **All** category. 
5. Scroll using the scrollbar (if you can see it). Also, enable touch simulation and scroll using the flick of your mouse. 
6. Select a category and confirm the page reloads with relevant results. 
7. Switch over to the **Themes** tab and confirm the category selector continues to work there. 
8. Switch over to the **Business Services** tab and confirm the category selector continues to work there. 
9. Test all tabs using a mobile viewport. This size doesn't have the scrollable category selector but just has the old dropdown as usual.
10. Delete or install the Woo Update Manager to see the margins look good with or without it. I don't have WUM but to get rid of the notice, I returned early [from `PluginInstallNotice`](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-admin/client/marketplace/components/woo-update-manager-plugin/plugin-install-notice.tsx#L21)
11. Load the My Subscriptions page and confirm it continues to load. 
12. Please test around this issue. 

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [x] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->
Update In-App Marketplace category selector

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

raicem committed:
Marketplace: update category selector

Remove the dropdown on the desktop view and show all items, even if
overflowing. Added helper buttons to scroll to the right to show more.

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

raicem commented on plugins/woocommerce-admin/client/marketplace/components/category-selector/category-selector.tsx:
```diff
@@ -131,50 +215,60 @@ export default function CategorySelector(
 			</div>
+			{ isOverflowing && (
+				<>
+					<button
+						onClick={ scrollToPrevCategories }
+						className="woocommerce-marketplace__category-navigation-button woocommerce-marketplace__category-navigation-button--prev"
+						hidden={ scrollPosition === 'start' }
+						aria-label="Scroll to previous categories"
+					>
+						<Icon icon="arrow-left-alt2" />
+					</button>
```
These buttons are overlayed on top of the category list to 

1) indicate there are more categories to see
2) help people scroll

In terms of accessibility and tab navigation, it's a bit clunky. I thought about removing them from the tab index but didn't act on it to see what you think

KokkieH replied:
See my main review comment...I agree they should be removed from the tab index. Keyboard nav through the categories work fine without using the buttons at all.

raicem replied:
Awesome, thank you for checking this! I have implemented it using `tabIndex={ -1 }`

raicem commented on plugins/woocommerce-admin/client/marketplace/components/products/products.tsx:
```diff
@@ -172,10 +172,12 @@ export default function Products( props: ProductsProps ) {
-				{ props.categorySelector && (
-					<CategorySelector type={ props.type } />
-				) }
+			<nav className="woocommerce-marketplace__sub-header">
```
I think the list of categories can be considered as a navigation. Just trying to avoid having too many divs

KokkieH replied:
I noticed VO reading the section out as a navigation area. This makes complete sense to me.

KokkieH commented on plugins/woocommerce-admin/client/marketplace/components/category-selector/category-selector.tsx:
```diff
@@ -72,21 +62,125 @@ export default function CategorySelector(
+		};
+	}, [ debouncedCheckOverflow, debouncedScrollPosition ] );
+
+	useEffect( () => {
+		checkOverflow();
+	}, [ categoriesToShow ] );
```
No change needed, but it might also work to call checkOverflow() in the first useEffect hook, after we call setCategoriesToShow(). Means we don't need another, separate useEffect hook here.

raicem replied:
I'd love to get rid of this useEffect but `setCategoriesToShow` doesn't immediately update the state. [Docs](https://react.dev/reference/react/useState#adding-state-to-a-component) say it as:

> It only affects what useState will return starting from the next render.

 So this may not work as there is a chance `checkOverflow` will be running too early

```js
setCategoriesToShow(categoriesToShow);
checkOverflow();
```

KokkieH replied:
Makes sense :)

KokkieH reviewed (approved):
Works as described.

Two non-blocking A11Y issues (I think it would be good if we can improve number 2, though)

Tested with keyboard nav and Voiceover as well and noticed a couple quirks:

1) If VO is on, the "Next" button doesn't disappear when I land on the last category (using tab-navigation). With VO off it does disappear. Not sure if this is something we can fix, though - likely a weird VO-browser interaction rather than a code issue.

2) After tabbing through all the categories, focus moves to the "Previous" button (and then the "Next" button, if it's visible) before going to the product cards. I'd argue in this case there's no need for the buttons to receive focus, as the next category automatically scrolls into view as focus tabs to it.

https://github.com/user-attachments/assets/70974a8d-d8c5-46ee-9334-289d92dff7f1

raicem committed:
Marketplace: remove category sroll helpers from tabindex

GitHub:
https://github.com/woocommerce/woocommerce/pull/51309/files#r1758448638

raicem commented:
Thank you so much for the review @KokkieH! 

I'll need to rebase this on recent search changes from @bor0. I'll rebase it either on top of the feature branch and merge it in or on top of PR [Remove in-app marketplace Search results tab and unify results into existing tabs #51297](https://github.com/woocommerce/woocommerce/pull/51297) and wait until it's merged.

raicem committed:
Marketplace: Remove selectedTab reference from product.tsx

This is probably included due to the merge conflict

raicem committed:
Marketplace: tweak category scroll button narrower

poligilad-auto commented:
Hey @raicem,
Just a question here, do we always see the scrollbar? Or is that only on mobile?

raicem commented:
> Just a question here, do we always see the scrollbar? Or is that only on mobile?

Hey @poligilad-auto, I don't think it's only mobile but it depends on some external factors (browser, user preferences, connected peripherals). I have Firefox on the left and Brave on the right. Brave has a very visible scroll bar but Firefox doesn't. 

[Screencast from 2024-09-17 20-59-30.webm](https://github.com/user-attachments/assets/9ea91799-be0e-4d68-9e3f-ac2c56e160b2)

Removing the scrollbar could impair some functionality but @mcliwanow pointed out that there might be ways to make it look much prettier, like the ones on [the Black Friday Trends page](https://woocommerce.com/black-friday-trends/#site-performance).

mcliwanow commented:
I was thinking later yesterday whether we need a scrollbar at all if we have arrows. @poligilad-auto, @raicem, another option would be to hide the scrollbar with CSS but still allow scrolling of the elements (while keeping the arrows).

poligilad-auto commented:
@raicem @mcliwanow How is WordPress.com doing it in `Plugins`? It seems to scroll and not show the scrollbar.

https://github.com/user-attachments/assets/aed6d75d-a937-4e8b-9b52-39060e28954a

Also pointing out that the background behind the arrow should be gradient enough (so that it seems to hide where is under it).

raicem commented:
@poligilad-auto thank you so much for providing the example, I didn't know it was possible to hide it but still have it scroll naturally (like on a touch device). I have implemented this and opened a PR. 

> Also pointing out that the background behind the arrow should be gradient enough (so that it seems to hide where is under it).

I think I matched the design with this one too, here is a screenshot. I compare the **Store Management** text fading away on the right end. 

![image](https://github.com/user-attachments/assets/93e80990-084f-46b1-807c-b65b68a3d2b2)

poligilad-auto commented:
Thanks @raicem!
Looking nice. I'll try to test it myself tomorrow if I have time.
We can always do smaller rounds of visual fixes later on.
